scripts/deploy/deploy-utils.ts

```diff
@@ -0,0 +1,121 @@
+  const commitMessages = execSync(
+    `git log ${currentlyDeployedCommit}..HEAD --oneline`,
+    { encoding: "utf-8" }
+  ).toString();
+  return commitMessages;
+};
+
+/**
+ * `getCurrentlyDeployedCommit` is a helper function that returns the currently deployed commit.
+ * It will call the `get-current-deployed-commit.sh` script, which will return either a git hash or a git tag.
+ * @returns - the currently deployed commit
+ */
+const getCurrentlyDeployedCommit = () => {
+  const currentlyDeployedCommit = execSync(
+    "bash scripts/deploy/get-current-deployed-commit.sh",
+    { encoding: "utf-8" }
+  )
+    .toString()
+    .trim();
+  return currentlyDeployedCommit;
+};
+
+/**
+ * `getCurrentCommit` is a helper function that returns the current commit.
+ * This is different from the currently deployed commit. The currently deployed commit is the commit that is currently deployed to production.
+ * The current commit is the commit that is currently checked out on your local machine and will be deployed to production.
+ * @returns - the current commit
+ */
+const getCurrentCommit = () => {
+  const currentCommit = execSync("git rev-parse HEAD", {
+    encoding: "utf-8",
+  })
+    .toString()
+    .trim();
+  return currentCommit;
+};
+
+/**
+ * `isOnMainBranch` is a helper function that checks if the current branch is the main branch.
+ * @returns true if the current branch is the main branch
+ */
+const isOnMainBranch = () => {
+  console.log("Checking if you are on the main branch");
+  const result = execSync("git branch --show-current", {
+    encoding: "utf-8",
+  });
+  const isOnMain = result.toString().trim() === "main";
+  if (!isOnMain) {
+    console.log("Currently on branch:", result.toString().trim());
+  }
+  return isOnMain;
+};
+
+/**
+ * `isWorkingDirectoryClean` is a helper function that checks if the working directory is clean (i.e. no uncommitted changes).
+ * @returns true if the working directory is clean
+ */
+const isWorkingDirectoryClean = () => {
+  const result = execSync("git status --porcelain", { encoding: "utf-8" });
+  return result.trim() === "";
+};
+
+const isRunningOnCI = () => process.env.CI === "true";
+
+const isDryRun = process.argv.includes("--dry-run");
+
+/**
+ * `runDeploy` is a helper function that actually performs the deploy.
+ * It builds the production bundle, deploys it to the production server, and sends an email.
+ */
+const runDeploy = () => {
+  console.log("GETTING CURRENTLY DEPLOYED COMMIT");
+  const currentlyDeployedCommit = getCurrentlyDeployedCommit();
+  console.log(currentlyDeployedCommit);
+  console.log("BUILDING");
+  execSync("yarn build:prod", { stdio: "inherit" });
+  console.log("BUILD COMPLETE");
+  console.log("SAVING CURRENT COMMIT in build/commit.txt");
+  const currentCommit = getCurrentCommit();
+  execSync(`echo ${currentCommit} > build/commit.txt`, {
+    stdio: "inherit",
+  });
+  console.log("DEPLOYING");
+  if (!isDryRun) {
```

Comment:
This approach feels kinda brittle to me since it's dependent on including a conditional in every util function. It would be pretty easy to forget this check in a new function or change and the whole concept would break.

Since the intent is primarily to confirm that the deploy is wired up correctly without calling any "impactful" functions, here's one idea: we could introduce a file that mocks each of these functions (and the tag utils) for a dry run. Then, when exporting the utils, if `isDryRun`, the mock utils can be exported; if not, export the actual functions. This would hopefully reduce the number of failure points. Open to other ideas too!